APE: An Updated Model for the Affiliated Package Ecosystem

eteq opened the pull request:
**Abstract:**
A revision of the concept of "affiliated packages" is proposed.  The idea of
affiliated packages has continued mostly unaltered since the original Vision
for Astropy.  However, time has revealed some potential improvements in how the
ecosystem is structured and the review process for affiliated packages.  Hence,
this APE proposes to make three changes:

* This APE creates the concept of "coordinated" packages - affiliated-like packages
  that are more closely overseen by the Astropy Project. While apparently
  significant, this change is mainly a recognition of de facto status that some
  affiliated packages already have.
* The APE changes the review process to a more explicit criteria-based review
  process - again, this is primarily simply formalizing a process that was
  already in place in the coordination committee.
* It lays out guidelines for the eventual goal of converting the
  affiliated package review process into a more "peer-review" style, where
  other members of the community can contribute to the review.

**Tentative discussion start/new comment deadline**: March 9, 2018

stargaser commented:
I think this sounds great overall.

A question for the coordinated packages: will major proposed changes to these have an APE-like process? The question has been raised specifically for astroquery. I suggest it would be good for all the coordinated packages.

keflavich commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+coordinating committee) overseeing who is maintaining the package and taking
+responsibility for continued maintenance in the event the original author(s) no
+longer can do so. This means maintainer roles will be created for all the
+coordinated packages, and have full status as astropy
```
remove 'full' here (again, a nitpick, but it implies there might be a 'partial' astropy role... I don't think there is?)

keflavich commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+concerns in the "Motivation" section above about the committee's review
+process. To counter these concerns, prior to this APE the coordination committee
+has informally adopted a checklist-based approach, checking if certain "bullet
+points" are met.  While the formal criteria are listed in relatively vague terms
```
again, no need for scare quotes

keflavich commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+has informally adopted a checklist-based approach, checking if certain "bullet
+points" are met.  While the formal criteria are listed in relatively vague terms
+on the affiliated package web site, the details of how those criteria are met
+have been fairly consistent using an unwritten but fairly standard list. In
```
fairly x2

keflavich commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+on the affiliated package web site, the details of how those criteria are met
+have been fairly consistent using an unwritten but fairly standard list. In
+practice, this has meant that affiliated packages are given "provisional" status
+if some of these items are checked off are met, or "full" if a larger set of
```
"are checked off are met" probably should just be "are met"

keflavich commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+==================================================
+
+On a longer term horizon, a criteria-based review checklist will allow the
+review process to change to a more "peer-review" model.  That is, instead of
```
scare quotes

keflavich commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+On a longer term horizon, a criteria-based review checklist will allow the
+review process to change to a more "peer-review" model.  That is, instead of
+every package being vetted by the coordination committee, the committee can act
+more as "editors", sometimes offering the review opportunity to other members of
```
""

bsipocz commented:
Huge 👍  to this long overdue update!

hamogu commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+the Astropy community.  While the details of this process will not be outlined
+here, as they by necessity should be flexible enough to accomodate the
+experiences of the first few reviewers, this APE sets this as the eventual
+*goal* for the affiliated package review process.
```
I think it would be useful to already outline this process here. It sounds as if this will essentially follow the format of established journal article peer review with (as the text says) the coordination committee as editors. So, why not include this in the APE? The fact that it might be implemented a little later is fine. In APEs we usually lay out interfaces, data structures etc. that are still going to be implemented at some point in the future. This almost sounds as we we want a second APE for that process. IT would be simpler to include the basics here and leave the rest to the editors.

hamogu commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+  `Ropensci <https://ropensci.org/>`_) that may have tools that can be re-used,
+  it is likely that setting up this process will take some time - this APE does
+  not outline the details of this implementation, but rather gives the broad
+  goals.
```
Do we need to be that complicated? I mean, mow many reviews are there? Taking the current list of affiliated packages it looks like about 1-2 per months. Can't we adopt the same process and leave the implementation to email and a repro that keeps all results that the editors want to open source (just paste reviews from emails into a text file)?
While a more general approach might be better in the long term, are we sure that it's worth the initial investment at this point? Astropy has done good work in building tools that are used routinely and sometimes outside of astropy, too (e.g. ci-helpers), but we've also occasional wasted quite some time overthinking a problem so that we now have a technically brilliant solution that took a long time to build and is now hard to maintain (e.g. the io.ascii readers) or had to be simplified later (the intii8al concepts for nddata). It sounds to me it might be easier to peer=review with a low entry cost solution (email) and see how it goes.

mhvk replied:
I agree with @hamogu here. Would suggest to explicitly write that an initial trial will be done simply using e-mail, with the editors responsible for putting the results up in a usable form. Then, after some experience, it can be decided whether something more complicated is useful, or whether the instructions to the referees need to be sharpened, etc.

bsipocz commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+First, the affiliated packages include a mix of two different kinds of packages.
+Some packages have been developed as a fairly integrated effort alongside the
+core package, meaning both feature planning assuming the existence of the
+affiliated packages, as well as significant overlap in the core development
```
but this would remove the fact that these packages are overseen by the same coordination committee rather than being the product of one single developer. And when needed new people can be put in charge of maintaining these packages while that's not really the case for a normal affiliated package.

mhvk replied:
Not really: according to the text later on, those people would become part of the astropy team. I just thought one should not suggest that one already has to be part of the astropy core team to have a core affiliated package (even though in practice this may usually be the case; it is really just a matter of feeling that the current text gives the impression of an exclusive club).

bsipocz replied:
ah, ok I agree that this shouldn't suggest being a closed "club" either. But difficult to see that one can "have a core affiliated package", as the nature of those package is that they come from the community need rather than applied for affiliated status by someone "owning" it.

eteq replied:
Yep, I see this point - I took @mhvk's wording and massaged it a bit.  Since me pushing up the new draft will zap this comment chain, maybe do another review if you have more suggestions on the new wording, @mhvk or @bsipocz ?

pllim commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+----------------------
+
+While in the future this APE proposes that new affiliated packages not be in the
+Astropy github organization, affiliated packages from *before* this APE that are
```
Nitpick: "github" -> "GitHub"

weaverba137 commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+  These criteria will need to be translated to HTML and put on the affiliated
+  package instructions section of the Astropy website.  The affiliated package
+  listing page will also need to be updated for these new criteria - a concept
+  for this new site is `viewable here <example_affilpkg_page.png>`_.
```
Could you remind us where the data are that was used to construct example_affilpkg_page.png?  Is that example meant to be representative of actual package status, or was it just created using random data?  For example, I'm pretty sure that PyDL is Python 3 compatible.

eteq replied:
I think the data for that example might have been *entirely* made-up.  That is, it's a mock up of what it *might* look like. The authoritative source for this is http://www.astropy.org/affiliated/registry.json , and you can see there that `PyDL` is "good" for Python3, so nothing to worry about there!

weaverba137 replied:
Good, thank you for checking.  Unless there are some privacy concerns that I am missing, maybe the best thing is actually to use the real data.  I think the graphics look good overall.  I'm guessing the title of the graph was also made up.

eteq replied:
The main problem is that this was a "mock up" and I believe it would take quite a bit of work to do a "real" version (and it wouldn't exercise all the features).  But maybe @granttremblay can confirm on this? (he mad the mockup a while back)

astrofrog replied:
Can we add a banner of text in the screenshot that says 'not actual data' or something like this?

eteq replied:
Done re: banner - hopefully that will also make it a bit clearer if anyone else looks and sees things like @weaverba137 did!

eteq commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+re-reviewed as a *whole* at some future date to become "full" affiliated
+packages. This "provisional" and "full" distinction is opaque and probably
+meaningless to the average user who just wants to know what a package does, and
+if it is any good. Additionally, while the review standards are stated on the
```
Actually, I agree with @mhvk's :-1: here - I think "is any good" is really what I meant by this: that is, the "average user" doesn't care nearly as much about "Community standards" as they do about if it is good *as they perceive it*.  We are trying to make the point that meeting community standards is the right kind of "good", but that's not actually the point they care most about.

eteq commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,212 @@
+Second, the current review process has become somewhat opaque and rather awkward
+to keep organized.  The coordination committee reviews the packages, which in
+practice nearly always receive a "provisional" designation, but then must be
+re-reviewed as a *whole* at some future date to become "full" affiliated
```
I'm about to push something out to address this - it leaves just one of them in (to introduce the concept of "provisional", but got rid of the rest)

eteq commented:
I think I addressed all the typo/minor inline comments in the commits I just pushed up.  I'm addressing the more general questions here, except for the @hamogu and @mhvk's suggestion of adding the peer review process (which I will wrap up into a separate comment since I think it's a distinct discussion)


From @stargaser :
> A question for the coordinated packages: will major proposed changes to these have an APE-like process? The question has been raised specifically for astroquery. I suggest it would be good for all the coordinated packages.

Good question.  The answer is basically "yes if they want to" (just like how APEs work in practice for the core).  But that's not at all obvious here.  I added a new paragraph (now third in the `Astropy "coordinated" and affiliated packages` section) describing this in more detail, including some words that the coordinated packages *generally* track the core package model.


From @keflavich :
> I have questions about the example affiliate package table. Is the format and/or documentation of that being discussed somewhere else? 

That *will* be discussed in a PR that follows once this APE is accepted (although it's already been discussed in the `astropy-dev` in the form of the google doc with the review categories).  The idea is that we do *not* want to be too prescriptive of the details in this APE so that we don't need to make another APE if later on we want to add a column or the like.

eteq commented:
@mhvk @hamogu - I think you're glossing over some details here: How do we determine *who* the reviewers are?  And do we want them to be single- or even double-blind?  And it already is a bit more complicated than just email because we (the coordination committee) have scripts to ensure the issues/comments the package authors get a consistent message with the relevant comments.  If the goal of this process is to write things down so others can do more of this to lower the lead time, the procedures for that need to be written down so that someone other than a CC member knows how to do it.

Of course we can (and should!) decide all that in stages, probably starting with the simplest possible thing exactly as you suggest.  But the point of this wording in the APE is that we do *not* want to write the details down in the APE because if we do then we need to have another APE to *change* them.  So the intent of not putting it in the text here is to make it more agile as we are experimenting with getting things up and running.

I can write some words laying this out more explicitly though, even saying something like "a likely starting point will be to begin by sending review opportunities to existing affiliated package authors or core developers"? Would that help, @mhvk  and @hamogu?  I just don't want to say a *definite* rule here because of my concern above.

mhvk commented:
@eteq - re-reading my comment, I think I say exactly what you suggest to put in (even if I hadn't suggested to put in my comment), so I'm fine with your suggestion.

hamogu commented:
@eteq Yes, that sounds good. I just don't like and APE that essentially says "the details of the process that we propose in this APE are left to another APE". A short sentence "this is how we'll do it initially" is enough at this point.

eteq commented:
@mhvk @hamogu - what do you think of the commit I just pushed up - sufficient?

mhvk commented:
Yes, very happy with that (note the error from using double back ticks, though).

hamogu commented:
Look good to me.
(And, just a note about your comment above: I very much doubt that a double-blind peer-review process can be implemented for any package that's openly developed on github where every commit is tagged with an author name and email. So, the eventual process is probably not going to be double blind).

eteq commented:
OK, great.  I fixed the typo @mhvk pointed out with the backticks.  So I think this I've addressed everything, and this is good for the coordination committee to give a 👍 / 👎 on!

Re @hamogu : 
> I very much doubt that a double-blind peer-review 

I think I agree too, although single-blind (anonymous reviewer) is a possibility.  I was mostly just illustrating that there are a lot of considerations here to call it all "done", and we wouldn't want to hash them out in this comment thread.

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+An Updated Model for the Affiliated Package Ecosystem
+-----------------------------------------------------
+
+author: Erik Tollerud, 2017 Astropy Coordination Meeting Attendees
```
@eteq - I would just list yourself - as long as you give credit to the coordination meeting attendees in the main text, you don't need to include them/us as authors. For one, this includes the coordination committee itself which makes the task of accepting this APE tricky, and second you were the actual author of the APE. In a few years people won't even know who *2017 Astropy Coordination Meeting Attendees* was - I don't think it makes sense to refer to vague collections of people in the author list.

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+Abstract
+--------
+
+A revision of the concept of "affiliated packages" is proposed.  The idea of
```
Here and elsewhere, I think this should say 'Astropy-affiliated packages' to be specific. We often say 'affiliated packages' colloquially, but the correct/formal term includes 'Astropy'. Or maybe after the first time you can say (hereafter 'affiliated packages')

eteq replied:
Good point, but I think I'll choose the second way just to lower the clutter.

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+
+A revision of the concept of "affiliated packages" is proposed.  The idea of
+affiliated packages has continued mostly unaltered since the original Vision
+for Astropy.  However, time has revealed some potential improvements in how the
```
Include a link to the vision? (in the astropy docs)

eteq replied:
It's also linked below, but it doesn't hurt to do it twice!

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+
+First, the affiliated packages include a mix of two different kinds of packages.
+Some packages have been developed as a fairly integrated effort alongside the
+core package.  That is, feature planning took into account that these
```
Here and elsewhere, remove double spaces after periods 😛

eteq replied:
:trollface: - but I'll do it! :wink:

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+process for more details of the current process in practice).
+
+These issues were discussed at length by the coordination committee, and later
+in the 2017 Astropy coordination meeting.  The new structure and processes
```
in -> at?

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+
+The biggest-picture change is to split the currently single-category affiliated
+package ecosystem into two categories: "Astropy coordinated" and "Astropy
+affiliated" package.  This change is not as drastic as it sounds because it
```
package -> packages

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+
+This also means the development process for coordinated packages will generally
+track close to the core package model - e.g. they will usually follow the
+directory structure of the core package, APEs can be used for major
```
'follow the directory structure' -> not clear what this means - maybe 'follow the same package layout' would be more accurate?

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+astropy organization can be "legacied-in".  This means that they may remain in
+the GitHub organization even if they are not coordinated packages. Moving
+forward, though, new coordinated packages should be in the astropy organization
+and new  affiliated packages should not be.
```
remove double space after 'new'

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+process. To counter these concerns, prior to this APE the coordination committee
+has informally adopted a checklist-based approach, checking if certain bullet
+points are met.  While the formal criteria are listed in relatively vague terms
+on the affiliated package web site, the details of how those criteria are met
```
web site -> page on the astropy website (and include link)

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+on the affiliated package web site, the details of how those criteria are met
+have been fairly consistent using an unwritten but roughly standard list. In
+practice, this has meant that affiliated packages are given "provisional" status
+if some of these items are met, or "full" if a larger set of them are.
```
Actually we've got rid of provisional status, so this shouldn't be mentioned?

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+"Implementation" section below), with various levels defined for each criteria,
+generally corresponding to "good", "acceptable", and "unacceptable". This
+will make the review process much faster and (eventually) open the review
+process beyond the coordination committee, thereby address the concerns from the
```
address -> addressing

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+* The new package review criteria have already been implemented by the
+  coordination committee and reviewed by the community (Available in Google doc
+  form `here <https://docs.google.com/document/d/15PJf2PROXMa7xwTDvWnjXB_9KNuO2Ia4_kkxJ7MPazE/edit?usp=sharing>`_.
+  These criteria will need to be translated to HTML and put on the affiliated
```
HTML may be too specific, we might want to put it in a JSON file (JSON ALL THE THINGS!)

eteq replied:
I meant that it needs to be stored in some way and *displayed* via HTML (even if it's DOM-modified via JS).  So I clarified that!

astrofrog commented on APE_coordinated_pkg.rst:
```diff
@@ -0,0 +1,235 @@
+  the coordination committee from the pool of existing affiliated package
+  authors or Astropy core team members.  They receive a prompt to review a new
+  package via email (along with the criteria described a above), and provide
+  their reviews also via email.  The coordination committee will then perform
```
reviews by email is too specific - maybe the reviews will be by GitHub. Maybe you could say that this is not a recommended starting point but an example? Or just say that the technological details are beyond the scope but the main thing is just the existence of reviewers as opposed to just the coordination committee.

eteq replied:
Good point.  Will be a bit less specific on that point.

eteq commented:
Alright, I think I've resolved @astrofrog's suggestions in uncontroversial manner.  If he agrees, I think this is ready for a final decision by the coordination committee!

astrofrog commented:
Looks good to me! 👍

taldcroft commented:
Accepted by the Astropy Coordination Committee on May 14, 2018.
